frontend/src/components/common/Terminal.tsx

```diff
@@ -116,24 +218,50 @@ export default function Terminal(props: TerminalProps) {
         return;
       }
 
-      const xterm = new XTerminal();
+      if (xtermRef.current) {
+        xtermRef.current.dispose();
+        execRef.current?.cancel();
+      }
+      xtermRef.current = new XTerminal({
+        cursorBlink: true,
+        cursorStyle: 'underline',
+        scrollback: 10000,
+        rows: 30, // initial rows before fit
+      });
+
+      fitAddonRef.current = new FitAddon();
+      xtermRef.current.loadAddon(fitAddonRef.current);
+      (xtermRef.current as any).connected = false;
 
-      const fitAddon = new FitAddon();
-      xterm.loadAddon(fitAddon);
+      xtermRef.current.writeln(t('Connecting…') + '\n');
 
-      xterm.writeln(t('Connecting…') + '\n');
+      (async function () {
+        execRef.current = await item.exec(
+          container,
+          shells.available[shells.currentIdx],
+          (items: ArrayBuffer) => onData(xtermRef.current, items)
+        );
 
-      const exec = item.exec(container, (items: ArrayBuffer) => onData(xterm, items));
+        setupTerminal(terminalContainerRef, xtermRef.current, fitAddonRef.current);
+      })();
 
-      setupTerminal(terminalContainerRef, xterm, fitAddon, exec);
+      const handler = () => {
+        fitAddonRef.current.fit();
+        console.log('resize');
+      };
+      window.addEventListener('resize', handler);
 
       return function cleanup() {
-        xterm.dispose();
-        exec.cancel();
+        console.log('[Terminal] run cleanup.');
+        xtermRef.current.dispose();
+        if (execRef.current) {
+          execRef.current.cancel();
+        }
+        window.removeEventListener('resize', handler);
       };
     },
     // eslint-disable-next-line react-hooks/exhaustive-deps
-    [container, terminalContainerRef]
+    [container, terminalContainerRef, shells.available[shells.currentIdx]]
```

Comment:
Should we use the plain `shells.available` instead of `shells.available[shells.currentIdx]`?